### PR TITLE
refactor: use iterators in seg hashtable

### PR DESCRIPTION
**Pull request by brayniac**
As noted in #414, we can use iterators to reduce code duplication
within the hashtable module.

This change adds two iterator types to the hashtable, allowing us
to iterate through item info entries for a given hash.

### CODE REVIEW

File: src/rust/storage/seg/src/hashtable/mod.rs [hunk 1]
```diff
@@ -109,6 +109,131 @@ pub(crate) struct HashTable {
...
+            return None;
+        }
+
+        let n_item_slot = if self.chain_idx == self.chain_len {
+            N_BUCKET_SLOT
+        } else {
+            N_BUCKET_SLOT - 1
+        };
+
+        let item_info = unsafe { &mut (*self.ptr.add(self.bucket_id)).data[self.item_slot] };
```
Comment (hderms) [inline src/rust/storage/seg/src/hashtable/mod.rs:217]: @brayniac was curious about race conditions w/ HashTable and whether there's some kind of concurrency safe machinery that makes use of unsafe here sensible. Is that accomplished on some higher level via the CAS-related functions I saw elsewhere in this file or through single-threaded execution or something? 

I could probably do better trying to answer this question for myself but I lack familiarity with Segcache

File: src/rust/storage/seg/src/hashtable/mod.rs [hunk 2]
```diff
@@ -109,6 +109,131 @@ pub(crate) struct HashTable {
...
+}
+
+struct IterMut<'a> {
+    ptr: *mut HashBucket,
+    bucket_id: usize,
+    item_slot: usize,
+    chain_len: usize,
+    chain_idx: usize,
+    finished: bool,
+    _marker: std::marker::PhantomData<&'a mut u64>,
```
Comment (hderms) [inline src/rust/storage/seg/src/hashtable/mod.rs:128]: maybe a comment for the intention behind the phantom data would be helpful for future readers

File: src/rust/storage/seg/src/hashtable/mod.rs (hunk 1 above)
Comment (brayniac) [reply src/rust/storage/seg/src/hashtable/mod.rs:217]: The implementation is not designed for concurrent access. We have the storage owned by a single thread. The cas function here allows the memcache cas command to be implemented and doesn't imply concurrency within the server. So it's really just cas for the clients and execution order is serialized in process.

File: src/rust/storage/seg/src/hashtable/mod.rs [hunk 3]
```diff
@@ -109,6 +109,131 @@ pub(crate) struct HashTable {
...
+impl<'a> Iter<'a> {
+    fn new(hashtable: &'a HashTable, hash: u64) -> Self {
+        let bucket_id = (hash & hashtable.mask) as usize;
+        let bucket = hashtable.data[bucket_id];
+        let chain_len = chain_len(bucket.data[0]) as usize;
+
+        Self {
+            hashtable,
+            bucket_id,
+            item_slot: 1,
```
Comment (hderms) [inline src/rust/storage/seg/src/hashtable/mod.rs:140]: yeah I came to a similar conclusion that we should just start on the 1st item as the 0th one is the bucket info, right?

File: src/rust/storage/seg/src/hashtable/mod.rs [hunk 4]
```diff
@@ -162,84 +287,48 @@ impl HashTable {
...
-            bucket = &mut self.data[bucket_id as usize];
         }
 
-        for chain_idx in 0..=chain_len {
-            let n_item_slot = if chain_idx == chain_len {
-                N_BUCKET_SLOT
-            } else {
-                N_BUCKET_SLOT - 1
-            };
+        let rand: u64 = self.rng.gen();
```
Comment (hderms) [inline src/rust/storage/seg/src/hashtable/mod.rs:302]: nice

**Comment by hderms** [comment]
huge improvement to readability, imo

**Comment by brayniac** [comment]
Reminder to myself to do some benchmarking

**Review by thinkingfish — APPROVED** [review]

File: src/rust/storage/seg/src/hashtable/mod.rs [hunk 5]
```diff
@@ -541,55 +520,35 @@ impl HashTable {
...
-                            } else {
-                                freq = (freq | 0x80) << FREQ_BIT_SHIFT;
-                            }
-                            if bucket.data[i] == current_info {
-                                bucket.data[i] = (current_info & !FREQ_MASK) | freq;
-                            }
-                        }
-
-                        if cas == get_cas(bucket.data[0]) {
-                            // TODO(bmartin): what is expected on overflow of the cas bits?
```
Comment (brayniac) [inline src/rust/storage/seg/src/hashtable/mod.rs:579]: Good catch @thinkingfish - Per discussion this is fixed as part of #413, comment should be removed

File: src/rust/storage/seg/src/hashtable/mod.rs [hunk 6]
```diff
@@ -541,55 +520,35 @@ impl HashTable {
...
                         }
+                        *item_info = (*item_info & !FREQ_MASK) | freq;
                     }
-                }
-            }
 
-            if chain_idx == chain_len {
-                break;
+                    if cas == get_cas(self.data[bucket_id as usize].data[0]) {
+                        // TODO(bmartin): what is expected on overflow of the cas bits?
```
Comment (brayniac) [inline src/rust/storage/seg/src/hashtable/mod.rs:544]: Same here

**Pull request merged**